aws/resource_aws_workspaces_workspace.go

```diff
@@ -0,0 +1,403 @@
+
+	tags := keyvaluetags.New(d.Get("tags").(map[string]interface{})).IgnoreAws().WorkspacesTags()
+
+	input := &workspaces.WorkspaceRequest{
+		BundleId:                    aws.String(d.Get("bundle_id").(string)),
+		DirectoryId:                 aws.String(d.Get("directory_id").(string)),
+		UserName:                    aws.String(d.Get("user_name").(string)),
+		RootVolumeEncryptionEnabled: aws.Bool(d.Get("root_volume_encryption_enabled").(bool)),
+		UserVolumeEncryptionEnabled: aws.Bool(d.Get("user_volume_encryption_enabled").(bool)),
+		Tags:                        tags,
+	}
+
+	if v, ok := d.GetOk("volume_encryption_key"); ok {
+		input.VolumeEncryptionKey = aws.String(v.(string))
+	}
+
+	input.WorkspaceProperties = expandWorkspaceProperties(d.Get("workspace_properties").(*schema.Set).List())
+
+	log.Printf("[DEBUG] Creating workspace...\n%#v\n", *input)
+	resp, err := conn.CreateWorkspaces(&workspaces.CreateWorkspacesInput{
+		Workspaces: []*workspaces.WorkspaceRequest{input},
+	})
+	if err != nil {
+		return err
+	}
+
+	wsFail := resp.FailedRequests
+
+	if len(wsFail) > 0 {
+		return fmt.Errorf("workspace creation failed: %s", *wsFail[0].ErrorMessage)
+	}
+
+	wsPending := resp.PendingRequests[0]
+
+	log.Printf("[DEBUG] Waiting for workspace to be available...")
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{
+			workspaces.WorkspaceStatePending,
+			workspaces.WorkspaceStateStarting,
+		},
+		Target:       []string{workspaces.WorkspaceStateAvailable},
+		Refresh:      workspaceRefreshStateFunc(conn, *wsPending.WorkspaceId),
+		PollInterval: 30 * time.Second,
+		Timeout:      25 * time.Minute,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("workspace %q was not available: %s", d.Id(), err)
+	}
+
+	wsAvailable := wsPending
+
+	d.SetId(*wsAvailable.WorkspaceId)
+	log.Printf("[DEBUG] Workspace %q is available", d.Id())
+
+	return resourceAwsWorkspacesWorkspaceRead(d, meta)
+}
+
+func resourceAwsWorkspacesWorkspaceRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	workspace, state, err := workspaceRefreshStateFunc(conn, d.Id())()
+	if err != nil {
+		return fmt.Errorf("error reading workspace (%s): %s", d.Id(), err)
+	}
+	if state == workspaces.WorkspaceStateTerminated {
+		log.Printf("[WARN] workspace (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	ws := workspace.(*workspaces.Workspace)
+	d.Set("bundle_id", ws.BundleId)
+	d.Set("directory_id", ws.DirectoryId)
+	d.Set("ip_address", ws.IpAddress)
+	d.Set("computer_name", ws.ComputerName)
+	d.Set("state", ws.State)
+	d.Set("root_volume_encryption_enabled", ws.RootVolumeEncryptionEnabled)
+	d.Set("user_name", ws.UserName)
+	d.Set("user_volume_encryption_enabled", ws.UserVolumeEncryptionEnabled)
+	d.Set("volume_encryption_key", ws.VolumeEncryptionKey)
+	if err := d.Set("workspace_properties", flattenWorkspaceProperties(ws.WorkspaceProperties)); err != nil {
```

Comment:
@gdavison What is wrong with `flattenWorkspaceProperties` here? I constantly receive failed test with:
```
--- FAIL: TestAccAwsWorkspacesWorkspace_createWithEmptyWorkspaceProperties (923.76s)
    testing.go:654: Step 0 error: Check failed: 5 errors occurred:
        	* Check 3/7 error: aws_workspaces_workspace.test: Attribute 'workspace_properties.0.compute_type_name' not found
        	* Check 4/7 error: aws_workspaces_workspace.test: Attribute 'workspace_properties.0.root_volume_size_gib' not found
        	* Check 5/7 error: aws_workspaces_workspace.test: Attribute 'workspace_properties.0.running_mode' not found
        	* Check 6/7 error: aws_workspaces_workspace.test: Attribute 'workspace_properties.0.running_mode_auto_stop_timeout_in_minutes' not found
        	* Check 7/7 error: aws_workspaces_workspace.test: Attribute 'workspace_properties.0.user_volume_size_gib' not found
```
Do I mislead the return type – `[]map[string]interface{}`?

Reply:
@Tensho It's probably because `workspace_properties` is declared as `schema.TypeSet`. Given that there will only be max. 1 of these blocks, declaring as `schema.TypeList` is better (and the test should pass).
You'll need to change `properties := d.Get("workspace_properties").(*schema.Set).List()` to `properties := d.Get("workspace_properties").([]interface{})`

Could you also please change the PR title to reflect the updated resource name?
Thanks.

Reply:
@ewbankkit OK, I'll rework `workspace_properties` to `TypeList`. TBH, I'm tired to fight with all this type-ish shit. Could anyone say me what is the good use case for `TypeSet` in this case? I'm totally lost with this it.

Reply:
@Tensho, in general a `TypeList` should be used when there are multiple items that should preserve order, and `TypeSet` should be used for multiple items where the order doesn't matter.
There's a special case to create a nested block where a `TypeList` with `MaxItems: 1` is used. I have to admit that it's kinda weird 😄

Reply:
>  I have to admit that it's kinda weird 😄

I'm fine now an can continue the work here 😄